### DISCUSSION TITLE
Soft Crypto Token storage customization for OCI Vault

### DISCUSSION THREAD

**Discussion by tedrextr** [Keyfactor/ejbca-ce#396] [Q&A]
Hi all experts,
I am looking for a least cost way of integrating EjbCA with Oracle cloud (akai OCI).
By default, the soft crypto token key materials are stored in CryptoTokenData table when mysql db is used.
My goal is to move them into a more secure storage, specially OCI Vault: 
https://docs.oracle.com/en-us/iaas/Content/KeyManagement/home.htm
After some research, it seems like like I need to implement com.keyfactor.util.keys.token.BaseCryptoToken
At least, Azure integration seem to be done that way.
Then how do I integrate it back to EbjCA itself?
Thank you for any help and reference

**Comment by primetomas** [comment]
You would need to implement a new Crypto Token, basically replicating the Azure Key Vault crypto token. After doing that including any test cases (like Azure Key Vault), you would create a pull request for EJBCA for us to look at it.

**Reply by tedrextr** [reply]
Thanks Thomas, I will try as you said.

**Reply by tedrextr** [reply]
Tomas, do you have any instructions on how to setup eclipse+widlfly+ejbca_latest dev env?
Having worked on intellij and maven lately, I could appreciate that.

**Comment by tedrextr** [comment]
Is https://doc.primekey.com/ejbca site working?

**Reply by primetomas** [reply]
Not yet. Many developers use IntelliJ, so if you are to that you can stick with that.

**Reply by tedrextr** [reply]
Do you know what is the corresponding github commit for these docker images in https://hub.docker.com/r/keyfactor/ejbca-ce ?
I have pulled the latest source version from https://github.com/Keyfactor/ejbca-ce, but it does not match the latest docker image.
Is there any special file that can serve as a clue for matching github commit?

**Reply by primetomas** [reply]
This file would be the version commit. 
https://github.com/Keyfactor/ejbca-ce/blob/main/src/internal.properties
There is no guarantee that the container matches the source version, they are updated independently of each other and may differ in exact code. But the differences should be small enough so it doesn't matter.

**Comment by tedrextr** [comment]
Hi Thomas,
Is there any plan/advancement in making EjbCA compliant with newer versions
of WildFly, like at least v28 ?
When trying to integrate EjbCA with OCI JDK in WildFly env, we are getting
conflict between Jersey JAX-RS that is used by OCI and RestEasy JAX-RS
built into WildFly 26
WildFly 26 seems to be the latest version that is known to be compatible
with EjbCA.

With WildFly 28 we don't experience the above issue.

Thanks
Ted

**Reply by primetomas** [reply]
See https://github.com/Keyfactor/ejbca-ce/discussions/569

**Comment by tedrextr** [comment]
thanks for info and quick response!
